### PR TITLE
Kpenny pact contract

### LINKED ISSUE #60: BountySwap Faucet
Must have an interface for people to claim kadenaswap coins

### PR DESCRIPTION
**Pull request by ggobugi27**
Added kpenny contract with hardcoded deadlines

To be added: 
- ~~admin-redeem feature (doesn't require user's sig)~~ Don't support admin-redeem feature. 
- deadline specifics and tests
- fungible-test

Added test: 
- upgradability
- fungible test
- reserve / read reservation history
- add liquidity / swap
- redeem 
  1. redeem with partially locked balance in pair account (did not remove-liquidity before final_deadline) 
  2. redeem  after removing liquidity
  3. redeem to a different account
- coin.precision test at redeem
- deadline test

closes #60

### CODE REVIEW

**Comment by fmelp** [comment]
mmm honestly i would not structure the redeem like that at all. I was thinking something more along the lines of a redeem-one function, then mapping that redeem-one in a redeem-all function that just does through the ledger table and does it for every user. Redeem-one send the user a `(coin.transfer 'kpenny-bank' user (/ kpenny-balance EX-RATE)`, then changes the balance to 0. All of this guarded by an ADMIN guard with same keys as contract governance.

**Comment by ggobugi27** [comment]
> mmm honestly i would not structure the redeem like that at all. I was thinking something more along the lines of a redeem-one function, then mapping that redeem-one in a redeem-all function that just does through the ledger table and does it for every user. Redeem-one send the user a `(coin.transfer 'kpenny-bank' user (/ kpenny-balance EX-RATE)`, then changes the balance to 0. All of this guarded by an ADMIN guard with same keys as contract governance.

I see. I was thinking that they were for users, but actually they aren't. I added the `redeem-all` function, but this will redeem any unremoved liquidity tokens with its guard as `swap.exchange` module guard as well. This will fail if anyone creates account with the pair account name on coin contract, so we'll need to figure out that case as well.

**Comment by fmelp** [comment]
thanks @ggobugi27 i think this is ready to merge and deploy on testnet. If @sirlensalot approves, I can do all the testing today. Also figured out an easy way to make the frontend for the kpenny buys

File: pact/kpenny/kpenny.pact [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+
+  (defconst MINIMUM_PRECISION:integer 14)
+
+  (defconst CONVERSION_RATE:decimal 1000000.0)
+
+  (defconst KPENNY_BANK:string 'kpenny-bank)
+
+  (defconst SWAP_DEADLINE (time "2021-03-15T00:00:00Z"))
+
+  (defconst FINAL_DEADLINE (time "2021-03-30T00:00:00Z"))
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:64]:
```suggestion
  (defconst FINAL_DEADLINE (time (read-msg "final-deadline")))
```

File: pact/kpenny/kpenny.pact [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+    )
+
+
+  (defun fund:string (account:string amount:decimal guard:guard)
+    (with-capability (CREDIT account)
+      (require-capability (FUND))
+        (credit account guard amount))
+  )
+
+  (defun convert:string (account:string amount-kda:decimal)
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:117]:
```suggestion
  (defun reserve:string (account:string amount-kda:decimal)
```

File: pact/kpenny/kpenny.pact [hunk 3]
```diff
@@ -0,0 +1,241 @@
+(namespace (read-msg 'ns))
+(module kpenny GOVERNANCE
+
+  (implements fungible-v2)
+  (use swap.fungible-util)
+  (use util.guards)
+
+  (defschema conversion
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:8]:
```suggestion
  (defschema reservation
```

File: pact/kpenny/kpenny.pact [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+    account:string
+    amount-kda:decimal
+    amount-kpenny:decimal)
+
+  (deftable conversions:{conversion})
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:13]:
```suggestion
  (deftable reservations:{reservation})
```

File: pact/kpenny/kpenny.pact [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    (let ((newbal (- managed requested)))
+      (enforce (>= newbal 0.0)
+        (format "TRANSFER exceeded for balance {}" [managed]))
+      newbal)
+  )
+
+  (defconst MINIMUM_PRECISION:integer 14)
+
+  (defconst CONVERSION_RATE:decimal 1000000.0)
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:58]:
```suggestion
  (defconst RESERVATION_RATE:decimal 1000000.0)
```

File: pact/kpenny/kpenny.pact [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+  (defcap GOVERNANCE ()
+    (enforce-guard
+      (keyset-ref-guard 'swap-ns-admin)))
+
+  (defcap DEBIT (sender:string)
+    (enforce-guard (at 'guard (read ledger sender))))
+
+  (defcap CREDIT (receiver:string) true)
+
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:29]: Let's add a `RESERVE @event`, see function comments

File: pact/kpenny/kpenny.pact [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+  (defun convert:string (account:string amount-kda:decimal)
+    (enforce-guard (before-date SWAP_DEADLINE))
+    (coin.transfer account KPENNY_BANK amount-kda)
+    (let
+      ( (tx-id (hash {"account": account, "amount": amount-kda, "salt": (at "block-time" (chain-data))}))
+        (amount-kpenny (* amount-kda CONVERSION_RATE))
+        (g (at 'guard (coin.details account)))
+      )
+      (insert conversions (format "{}-{}" [account, tx-id])
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:125]: I'm wondering if a RESERVE event will maybe be enough provenance. Then you can just hit the block explorer to resolve disputes. In any case we should definitely emit the event `(RESERVE account amount-kda)`

File: pact/kpenny/kpenny.pact [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+
+  (defun redeem-one:string (account:string)
+    (require-capability (REDEEM))
+    (with-read ledger account
+      { "balance" := amount-kpenny,
+        "guard" := g
+      }
+      (if (= amount-kpenny 0.0)
+        "Account's kpenny balance is 0.0"
+        (coin.transfer-create KPENNY_BANK account g (/ amount-kpenny CONVERSION_RATE))
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:143]: you'll need to check (and test in repl) KDA precision enforcement

File: pact/kpenny/kpenny.pact [hunk 9]
```diff
@@ -0,0 +1,241 @@
+(namespace (read-msg 'ns))
+(module kpenny GOVERNANCE
+
+  (implements fungible-v2)
+  (use swap.fungible-util)
+  (use util.guards)
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:6]: 👍

File: pact/kpenny/kpenny.repl [hunk 10]
```diff
@@ -0,0 +1,326 @@
+
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.repl:1]: let's add this to CI https://github.com/kadena-io/kadenaswap/actions/runs/531551264/workflow#L36-L47

File: pact/kpenny/kpenny.pact (hunk 8 above)
Comment (sirlensalot) [reply pact/kpenny/kpenny.pact:143]: We shouldn't do `transfer-create` according to current semantics of reservations, where you have the KDA account already, and the KDA account name is your KPenny name. By this logic we should only use `transfer` here

File: pact/kpenny/kpenny.repl [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+(expect "kda balance before redeem for unremoved liquidity token"
+  0.00000005 (coin.get-balance (at 'account (swap.exchange.get-pair swap.kpenny test.abc))))
+
+(expect "kda balance for kpenny-bank"
+  0.0 (coin.get-balance 'kpenny-bank))
+
+(rollback-tx)
+
+;;ERROR CASE TO FIX: Redeem fails if account name exists on coin account
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.repl:258]: See comment above, we shouldn't `transfer-create` in redeem

File: pact/kpenny/kpenny.repl [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+(env-chain-data { 'block-time: (time "2021-03-30T00:00:01Z" )})
+
+(expect-failure
+  "pair account name exists on coin contract with a different guard"
+  "account guards do not match"
+  (swap.kpenny.redeem-all))
+
+(rollback-tx)
+
+;;Deadline Test
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.repl:326]: Is this a TODO? Maybe we should make a coverage report in the PR desc (what cases get covered)

File: pact/kpenny/kpenny.pact (hunk 8 above)
Comment (ggobugi27) [reply pact/kpenny/kpenny.pact]: After discussion, I got rid of `redeem-all`/`redeem-one` and instead one `redeem` function that only the user can call to any account with provided guard.

File: pact/kpenny/kpenny.repl [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+(commit-tx)
+
+
+;;redeem without removing liquidity
+
+(begin-tx)
+
+(env-sigs [
+  { 'key: ""
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.repl:138]: This is an anti-pattern -- install the capability in the smart contract code.

File: pact/kpenny/kpenny.pact [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+    )
+  )
+
+  (defun redeem:string (account:string redeem-account:string redeem-guard:guard)
+    (with-capability (REDEEM account)
+      (with-read ledger account
+        { "balance" := amount-kpenny
+        }
+        (let ((amount-kda (floor (/ amount-kpenny RESERVATION_RATE) (coin.precision))))
+          (coin.transfer-create KPENNY_BANK redeem-account redeem-guard amount-kda)
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.pact:155]: Need to install cap here

File: pact/kpenny/kpenny.repl [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+    19.9
+    0.0
+    0.0
+    "Bob"
+    "Bob"
+    (read-keyset 'bob)
+)
+
+(env-sigs [
+  { 'key: ""
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.repl:216]: See above

File: pact/kpenny/kpenny.repl [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+;; precision test for redeem
+
+(begin-tx)
+(use swap.exchange)
+(env-data
+  { 'bob: ["bob"]
+  , 'alice: ["alice"]
+  })
+(env-sigs [
+  { 'key: ""
```
Comment (sirlensalot) [inline pact/kpenny/kpenny.repl:257]: See above

File: pact/kpenny/kpenny.repl (hunk 12 above)
Comment (sirlensalot) [reply pact/kpenny/kpenny.repl:316]: ?

**Review by sirlensalot — CHANGES_REQUESTED** [review]

File: pact/kpenny/kpenny.pact (hunk 14 above)
Comment (ggobugi27) [reply pact/kpenny/kpenny.pact:156]: Fixed it !

File: pact/kpenny/kpenny.repl (hunk 12 above)
Comment (ggobugi27) [reply pact/kpenny/kpenny.repl:304]: Restructuring repl file with cases. pushes to come in soon.

**Review by sirlensalot — APPROVED** [review]

**Pull request merged**